# Customer Data Analysis (with Python) <a class="anchor"  id="title"></a>

Joy Cheng <br>
2024-03-01 <br>
Understanding Customer and Purchase Behaviors

### 1. Problem Defining (ASK)

#### **1.1 Background**
This synthetic dataset offers information about each transaction within an online marketplace, including customer demographics such as gender and age, as well as purchase data. While it may not provide exhaustive customer details, it encompasses essential information crucial for analysis. Specifically designed to facilitate analytical and predictive purposes, it enables tasks such as trend analysis, examination of purchase behavior, market basket analysis, and customer churn prediction. By leveraging this dataset, companies and analysts can explore the dynamics of online consumer behavior, deriving valuable insights to inform strategic decision-making in e-commerce.

For this project, the main focus will be on performing Exploratory Data Analysis (EDA) from various angles, including customer demographics, products, sales timeline, etc. Additionally, based on the findings, business recommendations and marketing strategies will be provided.

#### **1.2 Potential stakeholders**
* CEO of the company
* Sales or account executives
* Marketing and product team

#### **1.3 Questions to address**
During the ask phase of data analysis, it is crucial to consider your stakeholders and determine what type of information may be most valuable to them. In real-world situations, I would suggest requesting a quick meeting with the stakeholders and relevant teams to solicit their advice and input on the matter. This approach would enable me to better comprehend the questions or challenges they've encountered and assist in refining the areas and topics I intend to explore further during the analysis.

For the purpose of this practice, I have formulated some questions that I believe would be beneficial for growing the business and gaining a better understanding of the customer pool: 

1. Are there any trends regarding customer's demographics (age/gender)?
2. Which time of the year or season tends to have the most sales? Any particular product that stands out?
3. How can we determine if a customer is high value(potential)?

### Data Preperation (PREPARE)

#### **2.1 Data source, integrity, and limitations**
The dataset, downloaded from Kaggle and offered by SHRIYASH JAGTAP, is licensed under the Open Data Commons Open Database License (ODbL). This license ensures that the dataset complies with open data principles, allowing for its unrestricted use, sharing, and modification. Furthermore, the dataset was last updated 5 months ago and has been awarded a usability score of 10.0 by Kaggle, indicating its currency and high credibility.

**About the data** 
* File name: "ecommerce_customer_data_large.csv"
* File size: 21.29 MB
* The dataset is synthetically generated using the Faker Python library and may not accurately represent real-world data.  
* It contains 13 columns with each row recording a completed transaction that includes customer information, demographics, purchase history, product information, payment mothod, return and churn statuses. Each vaiable(column) is explained in the metadata. 

**Data Limitations** 
* **Location/Country** - Information regarding the customer's location or the origin of the purchase is not provided. This would be valuable for understanding the geographical distribution of customers and identifying market opportunities in different locations.
* **Product ID/Name** - The information about what the product is for each transaction is unavailable.
* **Return details unknown** - No details regarding the quantity and price that the customer returned. 
* **Transaction ID unavailable** - Each row in the dataset represents a product transaction made by a customer. However, the absence of a transaction ID prevents us from determining whether the customer's transaction at the same time and day includes purchases of different product types. This information can only be obtained if there is a transaction ID for matching and verification. 
    * To address this concern, I began by grouping transactions based on `Customer ID` and `Purchase Date` to identify instances where a customer has purchased multiple product types on the same day. Then, I calculated the total price and quantity of products purchased on each day to compare it with the value provided in the `Total Purchase Amount` field. The result revealed that there is only one transaction per customer per day for each product type.




### Data Processing (PROCESS)

#### **3.1 Tools used for analysis**
For this project, I'll use Python, along with its Pandas and Plotly packages, to perform my analysis. Pandas streamlines the handling of 2-dimensional dataframes, structuring data akin to Excel sheets, making the process of data cleaning and manipulation more intuitive. Meanwhile, Plotly's diverse toolkit for creating customizable and interactive modern-style charts ensures a smooth and effortless data visualization experience.

#### **3.2 Data cleaning & manipulation**

In [1]:
# intall python packages
import numpy as np # linear algebra
import pandas as pd # data processing
import plotly.express as px # create interactive charts

In [2]:
# read file
df = pd.read_csv("/kaggle/input/e-commerce-customer-for-behavior-analysis/ecommerce_customer_data_large.csv")

df.info() # get info about count and datatypes
df.shape  # returns a tuple of total # of rows and columns
df.head() # preview first 4 rows of dataset

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 250000 entries, 0 to 249999
Data columns (total 13 columns):
 #   Column                 Non-Null Count   Dtype  
---  ------                 --------------   -----  
 0   Customer ID            250000 non-null  int64  
 1   Purchase Date          250000 non-null  object 
 2   Product Category       250000 non-null  object 
 3   Product Price          250000 non-null  int64  
 4   Quantity               250000 non-null  int64  
 5   Total Purchase Amount  250000 non-null  int64  
 6   Payment Method         250000 non-null  object 
 7   Customer Age           250000 non-null  int64  
 8   Returns                202618 non-null  float64
 9   Customer Name          250000 non-null  object 
 10  Age                    250000 non-null  int64  
 11  Gender                 250000 non-null  object 
 12  Churn                  250000 non-null  int64  
dtypes: float64(1), int64(7), object(5)
memory usage: 24.8+ MB


Customer ID        Purchase Date Product Category  Product Price  Quantity  \
0        44605  2023-05-03 21:30:02             Home            177         1   
1        44605  2021-05-16 13:57:44      Electronics            174         3   
2        44605  2020-07-13 06:16:57            Books            413         1   
3        44605  2023-01-17 13:14:36      Electronics            396         3   
4        44605  2021-05-01 11:29:27            Books            259         4   

   Total Purchase Amount Payment Method  Customer Age  Returns Customer Name  \
0                   2427         PayPal            31      1.0   John Rivera   
1                   2448         PayPal            31      1.0   John Rivera   
2                   2345    Credit Card            31      1.0   John Rivera   
3                    937           Cash            31      0.0   John Rivera   
4                   2598         PayPal            31      1.0   John Rivera   

   Age  Gender  Churn  
0   31  Female      0  
1   31  Female      0  
2   31  Female      0  
3   31  Female      0  
4   31  Female      0

In [3]:
# count distinct values in each column
df.nunique()

Customer ID               49661
Purchase Date            249728
Product Category              4
Product Price               491
Quantity                      5
Total Purchase Amount      5245
Payment Method                3
Customer Age                 53
Returns                       2
Customer Name             39878
Age                          53
Gender                        2
Churn                         2
dtype: int64

In [4]:
# check duplicates, returns all duplicated rows 
df[df.duplicated(keep=False)] 

Empty DataFrame
Columns: [Customer ID, Purchase Date, Product Category, Product Price, Quantity, Total Purchase Amount, Payment Method, Customer Age, Returns, Customer Name, Age, Gender, Churn]
Index: []

**Observations during the checking process:** 

* **Duplicates**: 
    * No duplicaded transactions were found. 
    * `Customer Age` and `Age` columns seem to contain identical data; one should be removed to avoid confusion later on.
* **Missing values**: 
    * Some missing values are present in the `Returns` column, where the count value (202,618) is less than the count of other variables (250,000). These empty cells will need to be replaced with 0 as the default, indicating that the customer has not returned the product yet. 
        * Another way for checking missing values can be using `df.isnull().sum()` to verfity and count the total number of missing values for each column.
* **Datatype**: 
    * The `Returns` data should be integers rather than floats since there are only 1s and 0s to indicate whether the purchase has been returned or not (1=returned, 0=no return). 
    * The `Purchase Date` data is currently in object type and will need to be converted into datetimes. This will help make the process of conducting time series calculations and plotting more user-friendly and efficient.
* **Errors**: 
    * The calculated total purchase amount (`Price` * `Quantity`) in each transaction appears to differ from the value provided in the `Total Purchase Amount` field. This inconsistency may indicate inaccuracies within the dataset.

After checking and identify all the issues, I then proceed to start cleaning and reformating my dataset. 

In [5]:
# Replace all NaN to 0 for Returns
# Convert dtype to integer
df['Returns'] = df['Returns'].fillna(0).astype(int)

#check non-null count & dtype
df['Returns'].info()

<class 'pandas.core.series.Series'>
RangeIndex: 250000 entries, 0 to 249999
Series name: Returns
Non-Null Count   Dtype
--------------   -----
250000 non-null  int64
dtypes: int64(1)
memory usage: 1.9 MB


In [6]:
# Convert `Purchase Date` dtype to Datetime
df['Purchase Date'] = pd.to_datetime(df['Purchase Date'])

# Split Year and Month into new columns
df['Purchase Year'] = df['Purchase Date'].dt.year
df['Purchase Month'] = df['Purchase Date'].dt.month_name()

df[['Purchase Date','Purchase Year','Purchase Month']]

Purchase Date  Purchase Year Purchase Month
0      2023-05-03 21:30:02           2023            May
1      2021-05-16 13:57:44           2021            May
2      2020-07-13 06:16:57           2020           July
3      2023-01-17 13:14:36           2023        January
4      2021-05-01 11:29:27           2021            May
...                    ...            ...            ...
249995 2023-01-24 12:32:18           2023        January
249996 2021-06-04 05:45:25           2021           June
249997 2022-11-10 17:11:57           2022       November
249998 2021-06-27 14:42:12           2021           June
249999 2020-09-07 05:12:19           2020      September

[250000 rows x 3 columns]

In [7]:
# Re-calculate total purchase amount
# Remove incorrect `Total Purchased Amount` column
df['Total Price'] = df['Product Price'] * df['Quantity']
#df = df.drop(['Total Purchase Amount'], axis=1)

# Re-order columns & exclude `Total Purchased Amount` and `Age`
cols = ['Customer ID','Customer Name','Purchase Date','Purchase Year','Purchase Month','Product Category','Product Price','Quantity','Total Price','Payment Method','Returns','Gender','Customer Age','Churn']
df = df[cols]
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 250000 entries, 0 to 249999
Data columns (total 14 columns):
 #   Column            Non-Null Count   Dtype         
---  ------            --------------   -----         
 0   Customer ID       250000 non-null  int64         
 1   Customer Name     250000 non-null  object        
 2   Purchase Date     250000 non-null  datetime64[ns]
 3   Purchase Year     250000 non-null  int32         
 4   Purchase Month    250000 non-null  object        
 5   Product Category  250000 non-null  object        
 6   Product Price     250000 non-null  int64         
 7   Quantity          250000 non-null  int64         
 8   Total Price       250000 non-null  int64         
 9   Payment Method    250000 non-null  object        
 10  Returns           250000 non-null  int64         
 11  Gender            250000 non-null  object        
 12  Customer Age      250000 non-null  int64         
 13  Churn             250000 non-null  int64         
dtypes: d

In [8]:
# Save cleaned dataset to a csv file or copy to a new dataframe
df_2 = df.copy()
#df.to_csv('ecommerce_data_cleaned.csv', index=False)

After checking and addressing all missing values, duplicates, and inconsistent formatting, I now have a clean dataset ready for further exploration and analysis. Before proceeding to the next phase of the project, I will save and create a copy of this cleaned and organized dataset in a new file or dataframe. This ensures that I can easily refer back to the cleaned dataset as needed in the future.

### Data Exploration & Visualization (ANALYZE & SHARE)

### **4.1 Basic data summary & stats**


In [9]:
# Get an overall stats of the data
df_2.describe(include='all')

Customer ID    Customer Name                  Purchase Date  \
count   250000.000000           250000                         250000   
unique            NaN            39878                            NaN   
top               NaN  Michael Johnson                            NaN   
freq              NaN              119                            NaN   
mean     25017.632092              NaN  2021-11-06 19:26:22.286192384   
min          1.000000              NaN            2020-01-01 00:07:26   
25%      12590.000000              NaN            2020-12-03 08:10:12   
50%      25011.000000              NaN     2021-11-06 07:39:08.500000   
75%      37441.250000              NaN  2022-10-11 01:21:56.249999872   
max      50000.000000              NaN            2023-09-13 18:42:49   
std      14412.515718              NaN                            NaN   

        Purchase Year Purchase Month Product Category  Product Price  \
count   250000.000000         250000           250000  250000.000000   
unique            NaN             12                4            NaN   
top               NaN         August      Electronics            NaN   
freq              NaN          23160            62630            NaN   
mean      2021.377136            NaN              NaN     254.742724   
min       2020.000000            NaN              NaN      10.000000   
25%       2020.000000            NaN              NaN     132.000000   
50%       2021.000000            NaN              NaN     255.000000   
75%       2022.000000            NaN              NaN     377.000000   
max       2023.000000            NaN              NaN     500.000000   
std          1.074712            NaN              NaN     141.738104   

             Quantity    Total Price Payment Method        Returns  Gender  \
count   250000.000000  250000.000000         250000  250000.000000  250000   
unique            NaN            NaN              3            NaN       2   
top               NaN            NaN    Credit Card            NaN    Male   
freq              NaN            NaN          83547            NaN  125676   
mean         3.004936     765.954660            NaN       0.405904     NaN   
min          1.000000      10.000000            NaN       0.000000     NaN   
25%          2.000000     288.000000            NaN       0.000000     NaN   
50%          3.000000     604.000000            NaN       0.000000     NaN   
75%          4.000000    1135.000000            NaN       1.000000     NaN   
max          5.000000    2500.000000            NaN       1.000000     NaN   
std          1.414737     593.494204            NaN       0.491067     NaN   

         Customer Age         Churn  
count   250000.000000  250000.00000  
unique            NaN           NaN  
top               NaN           NaN  
freq              NaN           NaN  
mean        43.798276       0.20052  
min         18.000000       0.00000  
25%         30.000000       0.00000  
50%         44.000000       0.00000  
75%         57.000000       0.00000  
max         70.000000       1.00000  
std         15.364915       0.40039

In [10]:
# Summarize object variables
df_2.describe(include=['object'])

Customer Name Purchase Month Product Category Payment Method  Gender
count            250000         250000           250000         250000  250000
unique            39878             12                4              3       2
top     Michael Johnson         August      Electronics    Credit Card    Male
freq                119          23160            62630          83547  125676

**Overview of the dataset:**

* There are a total of 49,661 unique customers in the dataset. This count assumes that customers with the same name but different customer IDs are considered as separate individuals.
* There are more male customers than females. 
* The majority of customers are around the age of 40. 
* Among the four product types, Electronics is the most popular. 
* Most customers prefer to pay via credit cards. 

### **4.2 Explore demographics - Gender & Age**

#### **I. Trends about Gender**
1. Gender distribution 
2. Purchase price distribution by gender
3. Total & average price spent by gender and year
4. Churn and returns by genger


In [11]:
df_2.head()

Customer ID Customer Name       Purchase Date  Purchase Year  \
0        44605   John Rivera 2023-05-03 21:30:02           2023   
1        44605   John Rivera 2021-05-16 13:57:44           2021   
2        44605   John Rivera 2020-07-13 06:16:57           2020   
3        44605   John Rivera 2023-01-17 13:14:36           2023   
4        44605   John Rivera 2021-05-01 11:29:27           2021   

  Purchase Month Product Category  Product Price  Quantity  Total Price  \
0            May             Home            177         1          177   
1            May      Electronics            174         3          522   
2           July            Books            413         1          413   
3        January      Electronics            396         3         1188   
4            May            Books            259         4         1036   

  Payment Method  Returns  Gender  Customer Age  Churn  
0         PayPal        1  Female            31      0  
1         PayPal        1  Female            31      0  
2    Credit Card        1  Female            31      0  
3           Cash        0  Female            31      0  
4         PayPal        1  Female            31      0

In [12]:
# Overview stats of Gender and total price spent
gender_summary = df_2.groupby('Gender').agg(total_count=('Customer ID','size'),
                                            min_amount=('Total Price', 'min'),
                                            max_amount=('Total Price', 'max'),
                                            median_amount=('Total Price','median'),
                                            total_amount=('Total Price','sum'),
                                            average_amount =('Total Price','mean'),
                                            total_returned=('Returns','sum'),
                                            return_rates=('Returns','mean'),
                                            total_churned=('Churn','sum'),
                                            churn_rates=('Churn','mean')).reset_index()

gender_summary

Gender  total_count  min_amount  max_amount  median_amount  total_amount  \
0  Female       124324          10        2500          605.0      95363311   
1    Male       125676          10        2500          604.0      96125354   

   average_amount  total_returned  return_rates  total_churned  churn_rates  
0      767.054720           50284      0.404459          25067     0.201626  
1      764.866434           51192      0.407333          25063     0.199426

In [13]:
## 1. Gender distribution 
# calculate count for each gender
gender_count = df_2['Gender'].value_counts()
# create pie chart
fig1 = px.pie(values=gender_count, names=gender_count.index,
              color = gender_count.index,
              color_discrete_map={'Female':'#FF6692', 'Male':'#3366CC'},
              title='1. Gender Distribution')
# adjust in-graph text
fig1.update_traces(textposition='inside', textinfo='text', 
                   texttemplate='%{label}<br>%{value}<br>(%{percent})')
# adjust chart size and remove legend
fig1.update_layout(title={'x': 0.5, 'y': 0.9},
                   width=400, height=400, showlegend=False)


## 2. Purchase price distribution by gender
# Calculate the percentage of total purchased amount for each gender
sum_amount = gender_summary['total_amount'].sum()
gender_summary['total_amount_percentage'] = (gender_summary['total_amount']/sum_amount*100).round(2)

# create a bar chart 
fig2 = px.bar(gender_summary, x='total_amount',y='Gender',
              orientation='h',
              color='Gender',
              color_discrete_map={'Female': '#FF6692', 'Male': '#3366CC'},
              title='2. Total Purchase Amount by Gender',
              text=gender_summary.apply(lambda x: f"${x['total_amount']:,}<br>({x['total_amount_percentage']:.2f}%)", axis=1),
              labels={'total_amount':'Total Purchased Amount ($)'})
fig2.update_layout(title={'x': 0.5, 'y': 0.9},
                  yaxis_title= None, 
                  width=550, height=350, showlegend=False)


## 3 Average price spent breakdown by year
# Calculate the overall average spending and prepare data for plotting
overall_avg_price = df_2['Total Price'].mean()
avg_price_gender_year = df_2.groupby(['Gender', 'Purchase Year'], as_index=False)['Total Price'].mean()

# Create box plot visualizing total price by gender and year, including overall average line
fig3 = px.box(df_2, x='Gender', y='Total Price', color='Purchase Year',
              title='3. Total Price Spent by Gender, Breakdown by Year',
              labels={'Total Price': 'Total Purchased Amount ($)'})

fig3.add_hline(y=overall_avg_price, line_dash="dash", line_color="#325A9B",
               annotation_text=f'Overall Average: ${overall_avg_price:.2f}',
               annotation_position="top",  # Adjusted for better visibility
               annotation_font_color="#325A9B")

fig3.update_layout(title={'x': 0.5, 'y': 0.9}, 
                   width=700, height=450,
                   xaxis_title=None)

## 4. Customer churn and returns by gender
# Convert churn and return rates to percentage
gender_summary['churn_rates'] *= 100
gender_summary['return_rates'] *= 100


fig4 = px.bar(gender_summary, x='Gender', y=['total_churned', 'total_returned'],
             barmode='group', 
             title='4. Customer Churn and Returns by Gender',
             text_auto='.4s',
             labels={'value': 'Number of Customers'},
             hover_data={'churn_rates': True, 'return_rates': True}) 

# Adjust data shown in the hover text
fig4.update_traces(hovertemplate="Gender: %{x}<br>"
                                 "Total: %{y:,.0f}<br>"
                                 "Churn Rate: %{customdata[0]}%<br>"
                                 "Return Rate: %{customdata[1]}%<extra></extra>")

# Add annotations for churn/return rates
for index, row in gender_summary.iterrows():
    fig4.add_annotation(x=row['Gender'], y=row['total_churned'],
                        text=f"{row['churn_rates']:.2f}%", 
                        showarrow=False, xanchor='right', yanchor='bottom', 
                        font=dict(color='blue', size=12))

for index, row in gender_summary.iterrows():
    fig4.add_annotation(x=row['Gender'], y=row['total_returned'],
                        text=f"{row['return_rates']:.2f}%",
                        showarrow=False, xanchor='left', yanchor='bottom', 
                        font=dict(color='red', size=12))
    

fig4.update_layout(title={'x': 0.5, 'y': 0.9},
                   width=600, height=500, 
                   xaxis_title= None,
                   legend_title=None,
                   legend=dict(orientation="h",yanchor="bottom",
                              y=1.02, xanchor="center",x=0.5))

# display graphs
fig1.show()
fig2.show()
fig3.show()
fig4.show()

/opt/conda/lib/python3.10/site-packages/plotly/express/_core.py:2065: FutureWarning:

When grouping with a length-1 list-like, you will need to pass a length-1 tuple to get_group in a future version of pandas. Pass `(name,)` instead of `name` to silence this warning.

/opt/conda/lib/python3.10/site-packages/plotly/express/_core.py:2065: FutureWarning:

When grouping with a length-1 list-like, you will need to pass a length-1 tuple to get_group in a future version of pandas. Pass `(name,)` instead of `name` to silence this warning.

/opt/conda/lib/python3.10/site-packages/plotly/express/_core.py:2065: FutureWarning:

When grouping with a length-1 list-like, you will need to pass a length-1 tuple to get_group in a future version of pandas. Pass `(name,)` instead of `name` to silence this warning.



> **Gender distribution and trends:**
> * Both male and female customers exhibit similar average spending, approximately $$756, suggesting a diverse customer base.
> * The average annual expenditure from 2020 to 2023 remains stable for both genders, hovering around $2,725, with males slightly outspending females by 0.11%.
> * The presence of outliers beyond the upper fence and the gap between average and median points in the box plots indicates the existence of outliers in the dataset.These outliers may require further investigation to understand if they represent rare scenarios, errors in data entry or calculation, or other factors, influencing the decision to exclude or explore them for additional insights.
> * Both churn rates and return rates are similar for both genders, with a slight variation—females shows a slighly higher churn rate, and males show a marginally higher return rate. However, both of these differences are minimal and do not necessarily indicate that one is more likley to churn or return than the other. Nevertheless, both churn and return rates are unusually high and evenly distributed across genders, which is atypical in consumer behavior patterns. It is advisable to investigate the causes of these trends further to determine if there are any inaccuracies or errors in the data.

#### **II. Trends about Age**
1. Age distribution
2. Compare age distribution by gender
3. Average price spent by age range
4. Churned rate by age group

First, start with defining the age range into <br>
`Age Range: 0-9, 10-19, 20-29, 30-39, 40-49, 50-59, 60-69, and 70+`

In [14]:
# Define bins for age groups
age_bins = [0,9,19,29,39,49,59,69,np.inf]
age_labels = ['0-9','10-19','20-29','30-39','40-49','50-59','60-69','70+']

# Add a new column and assign age group category to each transaction in df_2
df_2['Age Group'] = pd.cut(df_2['Customer Age'], bins=age_bins, labels=age_labels) 

# Check if `Age Group` is assigned correctly and added to the dataset
df_2[['Customer ID','Customer Age','Age Group']].sort_values('Customer Age', ascending=False)

#df_2[df_2['Age Group'].isnull()] # check any missing values

Customer ID  Customer Age Age Group
213450        27866            70       70+
76898         27353            70       70+
76896         27353            70       70+
76895         27353            70       70+
239854         5761            70       70+
...             ...           ...       ...
157044        44571            18     10-19
83330         12369            18     10-19
93626          8277            18     10-19
93625          8277            18     10-19
190060        45939            18     10-19

[250000 rows x 3 columns]

In [15]:
## 1. Age distribution by age
# Ensure the 'Age Group' in df_2 is categorized according to age_labels
# this will help to connect the data points in the correct order in line chart
df_2['Age Group'] = pd.Categorical(df_2['Age Group'], categories=age_labels, ordered=True)

# Count each age group
age_group_count = df_2['Age Group'].value_counts().sort_index().reset_index(name='Count')

# Calculate the percentage for each age group
age_group_count['Percentage'] = round(age_group_count['Count'] / age_group_count['Count'].sum()*100, 2)
age_group_count['Percentage Text'] = age_group_count['Percentage'].astype(str) + '%'

# Plot a line chart with bars
fig1 = px.line(age_group_count,x='Age Group', y='Count',
               title='1. Age Distribution',
               markers=True,
               text='Percentage Text',
               labels={'Count': 'Number of Customers'})

fig1.update_traces(textposition="top center") #adjust text position for percentages 
fig1.add_bar(x=age_group_count['Age Group'], y=age_group_count['Count'], text=age_group_count['Count'], textposition='inside', name='Count')
fig1.update_traces(textfont_color='white', selector=dict(type='bar'))
fig1.update_layout(title={'x': 0.5, 'y': 0.9},
                   xaxis_title= None,
                   yaxis_title= 'Count',
                   width=750, height=450, 
                   showlegend=False)


## 2. Age distrbution by gender
# count each age group by gender
age_group_gender_count = df_2[['Gender','Age Group']].value_counts().reset_index(name='Count')

fig2 = px.bar(age_group_gender_count, x='Age Group', y='Count', 
              color='Gender', barmode='group', 
              color_discrete_map={'Female': '#FF6692', 'Male': '#3366CC'},
              category_orders={'Age Group': age_labels},
              title='2. Age Distribution by Gender',
              text_auto=True)
fig2.update_layout(title={'x': 0.5, 'y': 0.9},
                   width=750, height=450,
                   xaxis_title= None,
                   legend_title=None,
                   legend=dict(orientation="h",yanchor="bottom",
                               y=1.02, xanchor="center",x=0.5))


## 3. Average price spent by age group
# calculate overall and average of total spending for each age group
overall_avg_price = df_2['Total Price'].mean()
avg_price_age_group = df_2.groupby(['Age Group'])['Total Price'].mean().reset_index(name='Avg Price')

fig3 = px.bar(avg_price_age_group, x= 'Age Group', y= 'Avg Price', 
              color = 'Avg Price',
              color_continuous_scale= "peach", 
              category_orders={'Age Group': age_labels},
              title='3. Average Price Spent by Age Group',
              text= [f"${value:.2f}" for value in avg_price_age_group['Avg Price']],
              hover_data={'Age Group':True, 'Avg Price':"${:.2f}"})

# Adjust data and format shown in the hover data
fig3.update_traces(hovertemplate="Age Group: %{x}<br>"
                                 "Average Spending: %{y:$,.2f}")

# Add a line to show the overall average for spending 
fig3.add_hline(y=overall_avg_price, line_dash="dash", line_color="#1CBE4F",
               annotation_text=f'Overall Average: ${overall_avg_price:.2f}',
               annotation_font_color="#1CBE4F",
               annotation_position="top left")

fig3.update_layout(title={'x': 0.5, 'y': 0.9},
                   width=700, height=450,
                   xaxis_title= None, 
                   yaxis_title= 'Average Spending ($)',
                   yaxis=dict(range=[500, 800],  # Set range from 400 to 800
                              tick0=500,  # Set starting tick at 400
                              dtick=50),  # Set interval between ticks
                   showlegend=False)


## 4. Customer churn and returns by age group
# Group by age group and churn/return status
age_group_churned_returned = df_2.groupby('Age Group').agg(total_churned=('Churn','sum'),
                                                           churn_rates=('Churn','mean'),
                                                           total_returned=('Returns','sum'),
                                                           return_rates=('Returns','mean')).fillna(0).reset_index()

# Calculate the total churned/returned count and churn/return rates 
age_group_churned_returned['churn_rates'] *= 100
age_group_churned_returned['return_rates'] *= 100


fig4 = px.bar(age_group_churned_returned, x='Age Group', y=['total_churned', 'total_returned'],
             barmode='group', 
             category_orders={'Age Group': age_labels},
             title='4. Customer Churn and Returns by Age Group',
             text_auto= '.3s' ,
             labels={'value': 'Number of Customers'},
             hover_data={'churn_rates': True, 'return_rates': True}) 

# Adjust data shown in the hover text
fig4.update_traces(hovertemplate="Age Group: %{x}<br>"
                                 "Total: %{y:,.0f}<br>"
                                 "Churn Rate: %{customdata[0]}%<br>"
                                 "Return Rate: %{customdata[1]}%<extra></extra>")

fig4.update_traces(textfont={'size': 11}, textposition='inside')

# Add annotations for churn rates
for index, row in age_group_churned_returned.iterrows():
    fig4.add_annotation(x=row['Age Group'], y=row['total_churned'],
                        text=f"{row['churn_rates']:.1f}%", 
                        showarrow=False, xanchor='right', yanchor='bottom', 
                        font=dict(color='blue', size=11))

# Add annotations for return rates
for index, row in age_group_churned_returned.iterrows():
    fig4.add_annotation(x=row['Age Group'], y=row['total_returned'],
                        text=f"{row['return_rates']:.1f}%",
                        showarrow=False, xanchor='left', yanchor='bottom', 
                        font=dict(color='red', size=11))

fig4.update_layout(title={'x': 0.5, 'y': 0.9},
                   width=800, height=500, bargap=0.1, 
                   xaxis_title= None,
                   legend_title=None,
                   legend=dict(orientation="h",yanchor="bottom",
                              y=1.02, xanchor="center",x=0.5))


# display graphs
fig1.show()
fig2.show()
fig3.show()
fig4.show()

/opt/conda/lib/python3.10/site-packages/plotly/express/_core.py:2065: FutureWarning:

When grouping with a length-1 list-like, you will need to pass a length-1 tuple to get_group in a future version of pandas. Pass `(name,)` instead of `name` to silence this warning.

/tmp/ipykernel_18/2167840934.py:51: FutureWarning:

The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.

/tmp/ipykernel_18/2167840934.py:83: FutureWarning:

The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.

/opt/conda/lib/python3.10/site-packages/plotly/express/_core.py:2065: FutureWarning:

When grouping with a length-1 list-like, you will need to pass a length-1 tuple to get_group in a

> **Distribution and trends by Age:**
> * The largest age group appears to be between 20-29 years old, closely followed by the 30-39 age group. However, there is a noticeable drop in the number of customers in the age groups of 70+ and 10-19.
> * Across genders, there seems to be a relatively similar distribution of age groups, with slight variations in the counts. Males outnumber females in almost all age groups except for the 20-29 age group.
> * An interesting observation is that the average spending tends to increase as age increases.
> * The age groups 20-29 and 60-69 appear to have higher churn and return rates, suggesting that customers within these age brackets are more likely to both return items and discontinue their membership. This trend could be influenced by specific needs or preferences prevalent in these age demographics. 
> * The graph also suggests a potential correlation between return rates and churn rates: as return rates increase, churn rates also tend to rise. However, further investigation is necessary to validate this observation conclusively.



### **4.3 Explore Trends About Products & Sales**
1. Product distribution by category 
2. Product category total count vs. returns
3. Product sales revenue by year
4. Product sales revenue by gender
5. Product sales revenue by age group

In [16]:
# To include stats abour product reutrns, 
# we'll need to calculate the Return_Quantity when Returns == 1  
# (assuming all items are returned altogether for each transaction)
df_2['Return Quantity'] = df_2.apply(lambda row: row['Quantity'] if row['Returns'] == 1 else 0, axis=1)

product_summary_table = df_2.groupby('Product Category').agg(Count=('Product Category', 'size'),
                                               Order_Quantity=('Quantity', 'sum'),
                                               Total_Revenue=('Total Price','sum'),
                                               Average_Revenue=('Total Price','mean'),
                                               Return_Count=('Return Quantity', 'sum')).reset_index()
product_summary_table

Product Category  Count  Order_Quantity  Total_Revenue  Average_Revenue  \
0            Books  62247          186608       47578138       764.344274   
1         Clothing  62581          188688       47977746       766.650357   
2      Electronics  62630          187861       47801925       763.243254   
3             Home  62542          188077       48130856       769.576541   

   Return_Count  
0         76551  
1         76279  
2         76278  
3         76336

In [17]:
# 1. Product Distribution by category
# Calculate the overall percentage for product quantity sold
total_order_quantity = product_summary_table['Order_Quantity'].sum()
product_summary_table['Overall_Percentage'] = (product_summary_table['Order_Quantity'] / total_order_quantity) * 100

# Calculate the percentage for return
product_summary_table['Return_Rate'] = round((product_summary_table['Return_Count'] / product_summary_table['Order_Quantity']) * 100, 2)

product_summary_table

Product Category  Count  Order_Quantity  Total_Revenue  Average_Revenue  \
0            Books  62247          186608       47578138       764.344274   
1         Clothing  62581          188688       47977746       766.650357   
2      Electronics  62630          187861       47801925       763.243254   
3             Home  62542          188077       48130856       769.576541   

   Return_Count  Overall_Percentage  Return_Rate  
0         76551           24.840196        41.02  
1         76279           25.117074        40.43  
2         76278           25.006989        40.60  
3         76336           25.035741        40.59

In [18]:
# 1. Product Distribution and Returns by category
# Set the category order for plotting colors
category_order = ["Electronics", "Clothing", "Home", "Books"]

fig1 = px.bar(product_summary_table, 
             x='Product Category', 
             y=['Order_Quantity', 'Return_Count'],
             barmode='overlay', opacity=0.7,
             color='Product Category',
             color_discrete_sequence=px.colors.qualitative.T10,
             pattern_shape = 'variable',
             category_orders={'Product Category': category_order},
             title='Product Category Distribution vs. Returns',
             labels={'value': 'Return & Total Order Count'})

# Add overall percentage and returns rate as text on the graph
for i, row in product_summary_table.iterrows():
    fig1.add_annotation(x=row['Product Category'], y=row['Order_Quantity']-15000, text=f"Total:{row['Order_Quantity']/ 1000:,.2f}K<br>({row['Overall_Percentage']:.2f}%)",
                       font=dict(color='black', size=12), showarrow=False)

for i, row in product_summary_table.iterrows():
    fig1.add_annotation(x=row['Product Category'], y=row['Return_Count']-15000, text=f"Returned:{row['Return_Count']/ 1000:,.2f}K<br>({row['Return_Rate']:.2f}%)",
                       font=dict(color='white', size=12), showarrow=False)
    

fig1.update_layout(title={'x': 0.5, 'y': 0.9},
                  xaxis_title= None,
                  width=700, height=500,
                  showlegend=False)



# 2. Sales Revenue by Product Category
total_revenue = df_2.groupby(['Product Category','Purchase Year'])['Total Price'].sum().reset_index()
total_revenue['Purchase Year'] = total_revenue['Purchase Year'].astype(str)

fig2 = px.bar(total_revenue, x='Product Category', y='Total Price',
              color='Purchase Year', barmode= 'group',
              color_discrete_sequence=px.colors.qualitative.Set2,
              category_orders={'Product Category': category_order},
              title = '2. Sales Revenue by Product Category and Year',
              labels={'Total Price': 'Total Price ($)'})
fig2.update_traces(texttemplate='%{y:$.4s}', textposition='inside', textangle=90)
fig2.update_layout(title={'x': 0.5, 'y': 0.9},
                  xaxis_title= None,
                  width=700, height=500)


# 3. Product revenue by Gender
product_revenue_gender = df_2.groupby(['Product Category','Gender'])['Total Price'].sum().reset_index(name='Total Revenue')

fig3 = px.bar(product_revenue_gender,x='Product Category', y='Total Revenue', 
              color='Gender',
              color_discrete_map={'Female': '#FF6692', 'Male': '#3366CC'},
              barmode='group', 
              title='3. Sales Revenue by Product Category & Gender',
              category_orders={'Product Category': category_order},
              labels={'Total Revenue': 'Total Revenue ($)'})

fig3.update_traces(texttemplate='%{y:$.4s}', textposition='inside', textangle=90)

fig3.update_layout(title={'x': 0.5, 'y': 0.9},
                  xaxis_title= None,
                  width=700, height=500,
                  legend_title=None,
                  legend=dict(orientation="h",yanchor="bottom",
                              y=1.02, xanchor="center",x=0.5))


# 4. Product revenue by Age Group
product_revenue_age = df_2.groupby(['Product Category','Age Group'])['Total Price'].sum().reset_index(name='Total Revenue')

fig4 = px.bar(product_revenue_age[product_revenue_age['Total Revenue'] > 0], #filter out empty bars
              x='Age Group', y='Total Revenue',
              color='Product Category', barmode= 'group',
              color_discrete_sequence=px.colors.qualitative.T10,
              category_orders={'Product Category': category_order},
              title = '4. Product Sales Revenue by Category & Age Group',
              labels={'Total Revenue': 'Total Revenue ($)'})

fig4.update_traces(texttemplate='%{y:$.3s}', textangle=90)

fig4.update_layout(title={'x': 0.5, 'y': 0.9},
                  xaxis_title= None,
                  width=800, height=500,
                  legend_title=None,
                  legend=dict(orientation="h",yanchor="bottom",
                              y=1.02, xanchor="center",x=0.5))


fig1.show()
fig2.show()
fig3.show()
fig4.show()

/opt/conda/lib/python3.10/site-packages/plotly/express/_core.py:2065: FutureWarning:

When grouping with a length-1 list-like, you will need to pass a length-1 tuple to get_group in a future version of pandas. Pass `(name,)` instead of `name` to silence this warning.

/opt/conda/lib/python3.10/site-packages/plotly/express/_core.py:2065: FutureWarning:

When grouping with a length-1 list-like, you will need to pass a length-1 tuple to get_group in a future version of pandas. Pass `(name,)` instead of `name` to silence this warning.

/tmp/ipykernel_18/1380884117.py:71: FutureWarning:

The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.

/opt/conda/lib/python3.10/site-packages/plotly/express/_core.py:2065: FutureWarning:

When grouping with a length-1 list-like, you will need to pass a length-1 tuple to get_group in a futu

> **Product distribution and sales trends:**
> * Clothing has the highest demand of all categories at 25.12% while Books has the highest return rate at 41%. 
> * The overall return rates for all product types appear abnormally high and even, which would merit further investigation into the underlying causes.
> * Home and Clothing consistently generate the highest sales revenue year over year, while Books register the lowest sales revenue among the categories.
> * It appears that customers aged 20-29 show a higher preference for Home goods, while those in the 40-49 age group tend to favor Clothing more. Other age groups show a more even distribution among all product types.

### **4.4 Explore Time Series Pattern**
1. Monthly count of orders by year
2. Compare current and previous annual sales revenue
3. Monthly return and churn rates over time

In [19]:
# Quick overview of yearly sales revenue
annual_sales_summary = df_2.groupby('Purchase Year')['Total Price'].agg(['min', 'max','mean','median','sum',])
annual_sales_summary

min   max        mean  median       sum
Purchase Year                                         
2020            10  2500  766.229627   604.0  51931213
2021            10  2500  765.467205   608.0  51594786
2022            10  2500  766.071850   600.0  51774966
2023            10  2500  766.088024   604.0  36187700

In [20]:
# Set month names in order and dtype to category
month_order = ['January', 'February', 'March', 'April', 'May', 'June', 
               'July', 'August', 'September', 'October', 'November', 'December']

df_2['Purchase Month'] = pd.Categorical(df_2['Purchase Month'], 
                                        categories=month_order, 
                                        ordered=True)

# Calculate the total and average of monthly sales revenue
monthly_sales = df_2.groupby(['Purchase Year', 'Purchase Month']).agg(Count=('Customer ID', 'size'),
                                                                      Total_Revenue=('Total Price', 'sum'),
                                                                      Avg_Revenue=('Total Price', 'mean')).reset_index()

/tmp/ipykernel_18/2149485385.py:10: FutureWarning:

The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.



In [21]:
## 1. Monthly count of orders, split by years
# Convert years to string for bar chart
monthly_sales['Purchase Year'] = monthly_sales['Purchase Year'].astype(str)

fig1 = px.bar(monthly_sales, x='Purchase Month', y='Count', 
               color='Purchase Year', facet_row='Purchase Year',
               title='1. Number of Orders per Month (by Year)', 
               labels={'Count': 'Order Count', 'Purchase Year': 'Year'},
               hover_data={'Count': ':,.0f'},
               text_auto=True)

fig1.update_layout(title={'x': 0.5, 'y': 0.9},
                  xaxis_title= None, 
                  width=650, height=600,
                  showlegend=False)

## 2. Compare total monthly sales
# Filter out data points where 'Total_Revenue' is 0
fig2 = px.area(monthly_sales[monthly_sales['Total_Revenue'] != 0], x='Purchase Month', y='Total_Revenue', 
              color='Purchase Year', facet_col='Purchase Year',
              color_discrete_sequence=px.colors.qualitative.Dark2,
              title='2. Total Monthly Sales Revenue (by Year)',
              labels={'Total_Revenue': 'Total Revenue ($)', 'Purchase Year': 'Year'},
              hover_data={'Total_Revenue': ':$,.0f'},
              markers=True)

fig2.update_yaxes(tickprefix="$")

fig2.update_layout(title={'x': 0.5, 'y': 0.9},
                  xaxis_title= None,
                  xaxis2_title= None,
                  xaxis3_title= None,
                  xaxis4_title= None,
                  width=800, height=350,
                  showlegend=False)

## 3. Compare average monthly sales
fig3 = px.line(monthly_sales, x='Purchase Month', y='Avg_Revenue', 
              color='Purchase Year', facet_col='Purchase Year',
              color_discrete_sequence=px.colors.qualitative.Set2,
              title='3. Average Monthly Sales Revenue (by Year)',
              labels={'Avg_Revenue': 'Average Revenue ($)', 'Purchase Year': 'Year'},
              hover_data={'Avg_Revenue': ':$,.0f'},
              markers=True)

fig3.update_yaxes(tickprefix="$")

fig3.update_layout(title={'x': 0.5, 'y': 0.9},
                  xaxis_title= None,
                  xaxis2_title= None,
                  xaxis3_title= None,
                  xaxis4_title= None,
                  width=800, height=350,
                  yaxis=dict(range=[700, 840]), 
                  showlegend=False)


## 4. Monthly return and churn rate by year    
df_return_churn = df_2[['Customer ID','Purchase Date','Returns','Churn']]
df_return_churn['Purchase Date'] = df_return_churn['Purchase Date'].dt.strftime('%Y-%m')
rates = df_return_churn.groupby('Purchase Date').agg(
    Return_Rate=('Returns', 'mean'), Churn_Rate=('Churn', 'mean')).reset_index()

# Convert rates to percentage
rates['Return_Rate'] *= 100
rates['Churn_Rate'] *= 100

fig4 = px.line(rates, x='Purchase Date', y=['Return_Rate','Churn_Rate'], 
              title='4. Monthly Return and Churn Rates Over time',
              markers=True,
              hover_data={'variable':':name'})


fig4.update_traces(hovertemplate="Type: %{customdata}<br>"
                                 "Purchase Date: %{x|%Y-%m}<br>"
                                 "Rate: %{y:.2f}%")

# Calculate the positions for annotations
max_return_rate = rates.loc[rates['Return_Rate'].idxmax()]
min_return_rate = rates.loc[rates['Return_Rate'].idxmin()]
max_churn_rate = rates.loc[rates['Churn_Rate'].idxmax()]
min_churn_rate = rates.loc[rates['Churn_Rate'].idxmin()]

annotations = [
    {'x': max_return_rate['Purchase Date'], 'y': max_return_rate['Return_Rate']+1, 
     'text': f"highest: {max_return_rate['Return_Rate']:.2f}%", 'showarrow': False},
    {'x': min_return_rate['Purchase Date'], 'y': min_return_rate['Return_Rate']-1, 
     'text': f"lowest: {min_return_rate['Return_Rate']:.2f}%", 'showarrow': False},
    {'x': max_churn_rate['Purchase Date'], 'y': max_churn_rate['Churn_Rate']+1, 
     'text': f"highest: {max_churn_rate['Churn_Rate']:.2f}%", 'showarrow': False},
    {'x': min_churn_rate['Purchase Date'], 'y': min_churn_rate['Churn_Rate']-1, 
     'text': f"lowest: {min_churn_rate['Churn_Rate']:.2f}%", 'showarrow': False} ]

for annotation in annotations:
    fig4.add_annotation(x=annotation['x'], y=annotation['y'], text=annotation['text'], 
                        showarrow=annotation['showarrow'], font=dict(size=11))

fig4.update_layout(title={'x': 0.5, 'y': 0.9},
                  xaxis_title= None,
                  yaxis_title= 'Rates (%)',
                  width=750, height=450,
                  legend_title=None,
                  legend=dict(orientation="h",yanchor="bottom",
                              y=1.02, xanchor="center",x=0.5))   


#display all charts
fig1.show()
fig2.show()
fig3.show()
fig4.show()

/tmp/ipykernel_18/2102373463.py:60: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy



> **Patterns about sales and return/churn over time:**
> * The final transaction date recorded in the dataset is September 13, 2023, which accounts for the notably lower total annual revenue compared to previous years.
> * The annual sales revenue across all years appears consistent with minor fluctuations, typically with February having the lowest total sales, likely due to having fewer days in the month. September seems to be a slower month in the year, experiencing a drop in revenue after August.
> * Although February consistently shows the lowest sales figures for all years, it does not have the lowest average monthly sales except for 2023.
> * The return rate remains relatively stable over time, hovering around 40%. The number seems high and should require further investigation into the type of products being returned as well as customer's satisfaction to help reduce the rate. 
> * The churn rate shows slight fluctuations but remains under 22% through the years. 
 

### **4.5 Customer segmentation & ranking**
Segment customers by customer's individual Customer Lifetime Value(CLV) based on purchase history into High, Medium, and low tiers. Using the formula below: 

**Customer Lifetime Value (CLV)** = Customer Value x Average Customer Lifespan <br>
> **Average Purchase Value** = Total Revenue / Number of Orders <br>
> **Average Purchase Frequency** = Total Number of Purchases per Customer / Number of Customers <br>
> **Average Customer Lifespan** = Sum of Customer Lifespans / Number of Customers <br>
> **Customer Value** = Average Purchase Value x Average Purchase Frequency <br>

Next, utilize the calculated CLV to help identify the most valuable customers and further explore and understand their purchasing behaviors by comparing:

1. Customer segment distribution
2. Total sales revenue
3. Monthly sales revenue (by year)
4. Age distributions
5. Churn and return rates

In [22]:
# Calculate each customer's CLV
# Start with grouping data by customer's ID and sum their total order num and purchase amount
customer_df = df_2.groupby(['Customer ID','Customer Name']).agg(
                order_count=('Customer ID', 'size'), 
                total_revenue=('Total Price', 'sum'),
                avg_purchase_value=('Total Price','mean'),
                return_num=('Returns', 'sum'),
                churned=('Churn','sum')).reset_index()

# Re-assign churned values into boolean 1,0 based on the condition
customer_df['churned'] = customer_df['churned'].apply(lambda x: 1 if x >= 1 else 0)
customer_df

Customer ID     Customer Name  order_count  total_revenue  \
0                1     Dominic Cline            3           5600   
1                2       Crystal Day            6           6459   
2                3      Joseph Perez            4           3613   
3                4        Wyatt Love            5           4339   
4                5   Shannon Hoffman            5           2263   
...            ...               ...          ...            ...   
49656        49996      Cindy Holden            7           6107   
49657        49997     Anita Gardner            2           1592   
49658        49998       Tracy Smith           10           8440   
49659        49999     Laura Burgess            6           4188   
49660        50000  Kenneth Lawrence            7           3871   

       avg_purchase_value  return_num  churned  
0             1866.666667           0        0  
1             1076.500000           4        0  
2              903.250000           0        0  
3              867.800000           3        0  
4              452.600000           3        0  
...                   ...         ...      ...  
49656          872.428571           2        0  
49657          796.000000           0        0  
49658          844.000000           6        1  
49659          698.000000           3        0  
49660          553.000000           3        0  

[49661 rows x 7 columns]

In [23]:
# Calculate average purchase frequency (APF) per customer
customer_df['avg_purchase_freq'] = customer_df['order_count'] / len(customer_df)

# Calculate the average customer lifespan
# Calculate the difference between each customer's the first and last purchase date
customer_lifespans = df_2.groupby('Customer ID')['Purchase Date'].apply(lambda x: (x.max() - x.min()).days)
avg_lifespan_days = round(customer_lifespans.mean(), 1)

print(f"Customer Average Lifespan: {avg_lifespan_days:.1f} days")

Customer Average Lifespan: 828.4 days


In [24]:
# Calculate customer value for each customer
customer_df['customer_value'] = customer_df['avg_purchase_value'] * customer_df['avg_purchase_freq']

# Then calculate the final Customer Lifetime Value (CLV)
customer_df['CLV'] = customer_df['customer_value'] * avg_lifespan_days

customer_df.head()

Customer ID    Customer Name  order_count  total_revenue  \
0            1    Dominic Cline            3           5600   
1            2      Crystal Day            6           6459   
2            3     Joseph Perez            4           3613   
3            4       Wyatt Love            5           4339   
4            5  Shannon Hoffman            5           2263   

   avg_purchase_value  return_num  churned  avg_purchase_freq  customer_value  \
0         1866.666667           0        0           0.000060        0.112765   
1         1076.500000           4        0           0.000121        0.130062   
2          903.250000           0        0           0.000081        0.072753   
3          867.800000           3        0           0.000101        0.087372   
4          452.600000           3        0           0.000101        0.045569   

          CLV  
0   93.414148  
1  107.743211  
2   60.268807  
3   72.379284  
4   37.749324

In [25]:
# Re-order columns and create a summary stat for CLV
customer_df = customer_df[['Customer ID','Customer Name','churned','return_num','order_count',
                           'total_revenue','avg_purchase_value','avg_purchase_freq',
                           'customer_value','CLV']]
customer_df.describe()

Customer ID       churned    return_num   order_count  total_revenue  \
count  49661.000000  49661.000000  49661.000000  49661.000000   49661.000000   
mean   24993.104488      0.199936      2.043374      5.034131    3855.916413   
std    14434.429306      0.399956      1.413121      2.199399    2144.105477   
min        1.000000      0.000000      0.000000      1.000000      11.000000   
25%    12494.000000      0.000000      1.000000      3.000000    2271.000000   
50%    24987.000000      0.000000      2.000000      5.000000    3584.000000   
75%    37492.000000      0.000000      3.000000      6.000000    5154.000000   
max    50000.000000      1.000000     10.000000     17.000000   17448.000000   

       avg_purchase_value  avg_purchase_freq  customer_value           CLV  
count        49661.000000       49661.000000    49661.000000  49661.000000  
mean           765.418793           0.000101        0.077645     64.320919  
std            299.340938           0.000044        0.043175     35.766033  
min             11.000000           0.000020        0.000222      0.183492  
25%            559.600000           0.000060        0.045730     37.882773  
50%            745.500000           0.000101        0.072169     59.785055  
75%            943.500000           0.000121        0.103784     85.974378  
max           2500.000000           0.000342        0.351342    291.051795

In [26]:
# Segment customers into High, Medium, and Low
# Define cutoff points by 25th and 75th percentile
clv_summary = customer_df['CLV'].describe()
high_cutoff = clv_summary['75%']  # Extract data point for 75th percentile
low_cutoff = clv_summary['25%']   # Extract data point for 25th percentile

# Assign segment labels based on CLV values
def assign_segment(clv):
    if clv >= high_cutoff:
        return 'High'
    elif clv >= low_cutoff:
        return 'Medium'
    else:
        return 'Low'

# Apply segment assignment to each row in the DataFrame
customer_df['Segment'] = customer_df['CLV'].apply(assign_segment)

customer_df.sort_values(by='CLV', ascending=False).head(10)

/tmp/ipykernel_18/3784360579.py:17: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy



Customer ID     Customer Name  churned  return_num  order_count  \
6310          6347       Lori Taylor        0           8           14   
32210        32420     Jeffrey Jones        0           4           12   
25164        25324   Jennifer Valdez        0           5           11   
5771          5802   Mary Cunningham        1           5           12   
40656        40930      Michael Neal        0           5           14   
41518        41800  Richard Stephens        0           2           14   
42155        42442     Jeffrey Brown        0           2           10   
22609        22752       Jill Reeves        0           6           10   
40593        40866  Christine Herman        0           7           12   
34287        34514       Stacy Meyer        0           5           13   

       total_revenue  avg_purchase_value  avg_purchase_freq  customer_value  \
6310           17448         1246.285714           0.000282        0.351342   
32210          15538         1294.833333           0.000242        0.312881   
25164          15184         1380.363636           0.000222        0.305753   
5771           15049         1254.083333           0.000242        0.303035   
40656          14948         1067.714286           0.000282        0.301001   
41518          14720         1051.428571           0.000282        0.296410   
42155          14420         1442.000000           0.000201        0.290369   
22609          14283         1428.300000           0.000201        0.287610   
40593          14207         1183.916667           0.000242        0.286080   
34287          14140         1087.692308           0.000262        0.284730   

              CLV Segment  
6310   291.051795    High  
32210  259.190898    High  
25164  253.285790    High  
5771   251.033841    High  
40656  249.349051    High  
41518  245.545760    High  
42155  240.541431    High  
22609  238.256120    High  
40593  236.988357    High  
34287  235.870724    High

Now that we have all the customer's CLV calculated and segments assigned, we can use them to find out who our top 10 high-value customers are. We'll use Pandas embeded SQL to perform the query:

In [27]:
# Use query to find out the top 10 high-value customers 
# exclude ones who has churned
from pandasql import sqldf

print('* Top 10 High-Value Customers (Current) *')
print('')
print(sqldf('''SELECT DISTINCT `Customer ID`, `Customer Name`, CLV
                 FROM customer_df 
                WHERE Segment = 'High' AND churned == 0
                ORDER BY CLV DESC
                LIMIT 10'''))

* Top 10 High-Value Customers (Current) *

   Customer ID     Customer Name         CLV
0         6347       Lori Taylor  291.051795
1        32420     Jeffrey Jones  259.190898
2        25324   Jennifer Valdez  253.285790
3        40930      Michael Neal  249.349051
4        41800  Richard Stephens  245.545760
5        42442     Jeffrey Brown  240.541431
6        22752       Jill Reeves  238.256120
7        40866  Christine Herman  236.988357
8        34514       Stacy Meyer  235.870724
9        41855     Gregory Green  233.919035


In [28]:
# Also apply segment-tiers to df_2 for later use in plotting
# use merge to join data based on `Customer ID` in customer_df
df_2 = df_2.merge(customer_df[['Customer ID','Segment']], on='Customer ID', how='left')

# Rename the 'Segment' column to avoid conflicts
df_2.rename(columns={'Segment':'Segment Tier'}, inplace=True)

# Check for any missing values after join
df_2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 250000 entries, 0 to 249999
Data columns (total 17 columns):
 #   Column            Non-Null Count   Dtype         
---  ------            --------------   -----         
 0   Customer ID       250000 non-null  int64         
 1   Customer Name     250000 non-null  object        
 2   Purchase Date     250000 non-null  datetime64[ns]
 3   Purchase Year     250000 non-null  int32         
 4   Purchase Month    250000 non-null  category      
 5   Product Category  250000 non-null  object        
 6   Product Price     250000 non-null  int64         
 7   Quantity          250000 non-null  int64         
 8   Total Price       250000 non-null  int64         
 9   Payment Method    250000 non-null  object        
 10  Returns           250000 non-null  int64         
 11  Gender            250000 non-null  object        
 12  Customer Age      250000 non-null  int64         
 13  Churn             250000 non-null  int64         
 14  Age 

In [29]:
# Create an overall summary table for segment
# Calculate the total count/order/sales/churn and averages 
segment_summary = customer_df.groupby('Segment').agg(Total_Count=('Segment', 'size'),
                                               Total_Order_Count=('order_count', 'sum'),
                                               Total_Revenue=('total_revenue','sum'),
                                               Average_Revenue=('total_revenue','mean'),
                                               Total_Churned=('churned', 'sum'),
                                               Churn_Rates=('churned','mean'),
                                               Total_Return_Num=('return_num','sum')).reset_index()
# Calculate return rates for each segment
segment_summary['Return_Rates'] = segment_summary['Total_Return_Num'] / segment_summary['Total_Order_Count']

segment_summary

Segment  Total_Count  Total_Order_Count  Total_Revenue  Average_Revenue  \
0    High        12416              90887       84189209      6780.703044   
1     Low        12415              36321       17362509      1398.510592   
2  Medium        24830             122792       89936947      3622.108216   

   Total_Churned  Churn_Rates  Total_Return_Num  Return_Rates  
0           2482     0.199903             36850      0.405449  
1           2436     0.196214             14663      0.403706  
2           5011     0.201812             49963      0.406891

In [30]:
## 1. Customer segment distribution
# Set the segment order for plotting colors
segment_order = ["Low", "Medium", "High"]

fig1 = px.pie(segment_summary, values='Total_Count', names='Segment',
              color = 'Segment',
              color_discrete_sequence=px.colors.qualitative.Dark24,
              category_orders={'Segment': segment_order},
              title='1. Customer Segment Distribution')
fig1.update_traces(textposition='inside', textinfo='text', 
                   texttemplate='%{label}<br>%{value}<br>(%{percent})')
fig1.update_layout(title={'x': 0.5, 'y': 0.9},
                   width=400, height=400, showlegend=False)


## 2. Total purchase amount over time by segment
# Sum all revenues to calculate the spending percentage for each segment
sum_overall_revenue = segment_summary['Total_Revenue'].sum()
segment_summary['Revenue Percentage'] = (segment_summary['Total_Revenue'] / sum_overall_revenue) * 100

fig2 = px.bar(segment_summary, x='Segment', y='Total_Revenue', 
              color='Segment', 
              color_discrete_sequence=px.colors.qualitative.Dark24,
              title='2. Total Purchase Amount by Segment',
              labels={'Total_Revenue': 'Total Money Spent ($)'},
              category_orders={'Segment': segment_order},
              text=segment_summary.apply(lambda x: f"${x['Total_Revenue']:,}<br>({x['Revenue Percentage']:.2f}%)", axis=1),
              hover_data={'Total_Revenue': ':$,.0f'})
fig2.update_layout(title={'x': 0.5, 'y': 0.9},
                   width=550, height=450, showlegend=False)


## 3. Monthly Sales Revenue Over Time
# Convert `Purchase Date` to Year-Month format
# Calculate the total price by date and segment
segment_df = df_2[['Customer ID','Purchase Date','Total Price','Segment Tier']]
segment_df['Purchase Date'] = segment_df['Purchase Date'].dt.strftime('%Y-%m')
segment_monthly_sales = segment_df.groupby(['Purchase Date','Segment Tier'])['Total Price'].sum().reset_index(name='Sum')

fig3 = px.area(segment_monthly_sales, x='Purchase Date', y='Sum', 
              color='Segment Tier', 
              color_discrete_map={"Low":'#2E91E5', "Medium":'#E15F99', "High":'#1CA71C'},
              title='3. Monthly Sales Revenue Over Time by Segment',
              markers=True,
              symbol='Segment Tier',
              hover_data={'Segment Tier':':name','Purchase Date':True,'Sum':True})

fig3.update_traces(hovertemplate="Value: %{customdata}<br>"
                                "Purchase Date: %{x|%Y-%m}<br>"
                                "Sum: %{y:$,.3s}")

fig3.update_layout(title={'x': 0.5, 'y': 0.9},
                  xaxis_title= None,
                  yaxis_title= 'Total Revenue ($)',
                  width=750, height=500,
                  legend_title=None,
                  legend=dict(orientation="h",yanchor="bottom",
                              y=1.02, xanchor="center",x=0.5)) 


# 4. Age Distrbution by Segmtent
# Calculate the count for each age and segment group
segment_age_count = df_2.groupby(['Age Group','Segment Tier']).size().reset_index(name='Count')

fig4 = px.bar(segment_age_count, x='Age Group', y='Count', 
             color='Segment Tier', barmode='group',
             color_discrete_sequence=px.colors.qualitative.Dark24,
             category_orders={'Segment Tier': segment_order,'Age Group': age_labels},
             text_auto=True,
             title='4. Age Distribution by Segment')
fig4.update_layout(title={'x': 0.5, 'y': 0.9},
                   width=800, height=500)


## 5. Customer churn and returns by Segment
# Convert churn/return rates to percentage 
segment_summary['Churn_Rates'] *= 100
segment_summary['Return_Rates'] *= 100

fig5 = px.bar(segment_summary, x='Segment', y=['Total_Churned', 'Total_Return_Num'],
             barmode='group', 
             category_orders={'Segment': segment_order},
             title='5. Customer Churn and Returns by Segment',
             text_auto= '.3s' ,
             labels={'value': 'Number of Customers'},
             hover_data={'Churn_Rates': True, 'Return_Rates': True}) 

# Adjust data shown in the hover text
fig5.update_traces(hovertemplate="Segment Tier: %{x}<br>"
                                 "Total: %{y:,.0f}<br>"
                                 "Churn Rate: %{customdata[0]}%<br>"
                                 "Return Rate: %{customdata[1]}%<extra></extra>")

fig5.update_traces(textfont={'size': 11}, textposition='inside')

# Add annotations for churn rates
for index, row in segment_summary.iterrows():
    fig5.add_annotation(x=row['Segment'], y=row['Total_Churned'],
                        text=f"{row['Churn_Rates']:.2f}%", 
                        showarrow=False, xanchor='right', yanchor='bottom', 
                        font=dict(color='blue', size=11))

# Add annotations for return rates
for index, row in segment_summary.iterrows():
    fig5.add_annotation(x=row['Segment'], y=row['Total_Return_Num'],
                        text=f"{row['Return_Rates']:.2f}%",
                        showarrow=False, xanchor='left', yanchor='bottom', 
                        font=dict(color='red', size=11))

fig5.update_layout(title={'x': 0.5, 'y': 0.9},
                   width=600, height=500, 
                   xaxis_title= None,
                   legend_title=None,
                   legend=dict(orientation="h",yanchor="bottom",
                              y=1.02, xanchor="center",x=0.5))


###
fig1.show()
fig2.show()
fig3.show()
fig4.show()
fig5.show()

/opt/conda/lib/python3.10/site-packages/plotly/express/_core.py:2065: FutureWarning:

When grouping with a length-1 list-like, you will need to pass a length-1 tuple to get_group in a future version of pandas. Pass `(name,)` instead of `name` to silence this warning.

/tmp/ipykernel_18/1005251683.py:37: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy

/tmp/ipykernel_18/1005251683.py:63: FutureWarning:

The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.

/opt/conda/lib/python3.10/site-packages/plotly/express/_core.py:2065: FutureWarning:

When grouping with a length-1 list-like, you 

> **Insights discovered from customer segmentation:**
> * The medium segment holds the majority of customers, comprising 50% of the customer base, and contributes significantly to the total purchase amount, accounting for 47% of the total revenue, which amounts to 89 million dollars. It is closely followed by the high segment, contributing 44% of the total revenue, equivalent to 84 million dollars.
> * There appears to be stability and consistency in revenue generated by each segment over time. This suggests a reliable customer base within each segment. All three segments exhibit similar revenue patterns and fluctuations, yet the medium segment generally generates higher revenue whereas the low segment generates the least revenue at all times. 
> * The age distribution across segments is notably similar, with the majority of customers falling within the 20-69 age range. Specifically, the age group of 20-29 shows the highest concentration across all segments.
> * Churn rates and return rates across all three segments hover around 20% and 40%, respectively, with the medium segment exhibiting the highest churn and return rates. This indicates similarity in customer retention and return challenges across segments.

### **4.6 Payment Analysis**
1. Payment method distribution vs. returns
2. Payment method distribution by gender and age group
3. Payment method distribution by segment


In [31]:
payment_count_returns = df_2.groupby('Payment Method').agg(total_count=('Customer ID','size'),
                                                     total_return=('Returns','sum'),
                                                     return_rates=('Returns','mean')).reset_index()

payment_count_returns['percentage'] = payment_count_returns['total_count']/payment_count_returns['total_count'].sum()*100
payment_count_returns['return_rates'] *= 100

payment_count_returns[['Payment Method','total_count','percentage','total_return','return_rates']]

Payment Method  total_count  percentage  total_return  return_rates
0           Cash        83012     33.2048         33686     40.579675
1    Credit Card        83547     33.4188         33858     40.525692
2         PayPal        83441     33.3764         33932     40.665860

In [32]:
## 1. Payment method distribution vs. returns
payment_method_order = ['Cash', 'Credit Card','PayPal']

fig1 = px.bar(payment_count_returns, x='Payment Method', y=['total_count', 'total_return'],
             barmode='overlay', opacity=0.7,
             color='Payment Method',
             color_discrete_sequence=px.colors.qualitative.Prism,
             pattern_shape = 'variable',
             category_orders={'Payment Method': payment_method_order},
             title='1. Payment Method Distribution vs. Returns',
             labels={'value': 'Return & Total Order Count'})


# Add overall percentage and returns rate as text on the graph
for i, row in payment_count_returns.iterrows():
    fig1.add_annotation(x=row['Payment Method'], y=row['total_count']-5000, text=f"Total:{row['total_count']/ 1000:,.2f}K<br>({row['percentage']:.2f}%)",
                       font=dict(color='black', size=12), showarrow=False)

for i, row in payment_count_returns.iterrows():
    fig1.add_annotation(x=row['Payment Method'], y=row['total_return']-5000, text=f"Returned:{row['total_return']/ 1000:,.2f}K<br>({row['return_rates']:.2f}%)",
                       font=dict(color='white', size=12), showarrow=False)
    
fig1.update_traces(hovertemplate="Payment Method: %{x}<br>"
                                 "Count: %{y:,.0f}<extra></extra>")

fig1.update_layout(title={'x': 0.5, 'y': 0.9},
                  xaxis_title= None,
                  width=600, height=550,
                  showlegend=False)


## 2. Payment Method by gender and age group
payment_gender_age = df_2.groupby(['Payment Method','Gender','Age Group']).agg(Count=('Customer ID','size')).reset_index()

fig2 = px.bar(payment_gender_age, x='Age Group', y='Count', barmode='group',
              facet_col='Payment Method', #facet_row='Gender',
              color='Gender',
              color_discrete_map={'Female': '#FF6692', 'Male': '#3366CC'},
              category_orders={'Payment Method': payment_method_order},
              title='2. Payment Method Distribution by Gender and Age',
              labels={'Count': 'Number of Customers'},
              text_auto=True)

fig2.update_traces(width=0.5)

fig2.update_layout(title={'x': 0.5, 'y': 0.95},
                   bargap=0.2,
                   xaxis_title= None,
                   xaxis2_title= None,
                   xaxis3_title= None,
                   width=900, height=400,
                   legend_title=None,
                   legend=dict(orientation="h",yanchor="bottom",
                              y=1.14, xanchor="center",x=0.5))


## 3. Payment Method by Segment
payment_segment = df_2.groupby(['Payment Method','Segment Tier']).agg(Count=('Customer ID','size')).reset_index()
payment_segment['Percentage'] = payment_segment['Count']/payment_segment['Count'].sum()

fig3 = px.bar(payment_segment, x='Payment Method', y='Percentage', barmode='group',
              color='Segment Tier', #facet_col='Payment Method',
              color_discrete_sequence=px.colors.qualitative.Dark24,
              category_orders={'Payment Method': payment_method_order, 'Segment Tier': segment_order},
              title='3. Payment Method Distribution by Segment',
              labels={'Percentage': 'Distribution (%)'},
              text_auto='.1%')

fig3.update_layout(title={'x': 0.5, 'y': 0.9},
                   bargap=0.1,
                   xaxis_title= None,
                   width=600, height=450)


###
fig1.show()
fig2.show()
fig3.show()

/tmp/ipykernel_18/3627307101.py:33: FutureWarning:

The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.

/opt/conda/lib/python3.10/site-packages/plotly/express/_core.py:2065: FutureWarning:

When grouping with a length-1 list-like, you will need to pass a length-1 tuple to get_group in a future version of pandas. Pass `(name,)` instead of `name` to silence this warning.



> **Insights from payment analysis:**
> * Credit Card emerges as the most preferred payment method, while PayPal exhibits the highest return rate among the three.
> * Customer payment behavior patterns remain largely consistent across all age ranges and genders, with some variations observed: PayPal shows slightly higher demand among the 20-29 age group, and credit card and cash usage are more popular among females in their 30s.
> * Payment method distributions appears consistent across all segments, with minor variations where the low segment shows slightly higher usage of Cash and the high segment exhibits a slightly higher preference for PayPal. 

### **4.7 Find correlations**

In [33]:
## 1. Explore correlations between each quantitative variable
# round p-values to 3 decimals
df2_corr = round(df_2[['Product Price','Quantity','Total Price',
                       'Customer Age','Returns', 'Churn']].corr(method='pearson'), 3)
# Create a correlation heatmap
fig = px.imshow(df2_corr, 
                color_continuous_scale = 'blugrn',
                text_auto=True, aspect="auto")

fig.update_layout(title={'x': 0.5, 'y': 0.9},
                  width=800, height=450)    

fig.show()

> **Insights discovered from the correlation metrix:** 
<br>
> * **Product Price & Total Price**: There is a strong positive correlation (0.72) between Product Price and Total Price. This suggests that as the price of individual products increases, the total price of purchases also tends to increase, which is an expected relationship.
<br>
> *  **Quantity & Total Price**: There's also a moderately strong positive correlation (0.608) between the Quantity of items purchased and the Total Price. This indicates that as more items are purchased, the total price increases.
> *  **Customer Age, Returns, and Churn**: There's no strong linear relationships with other factors, meaning they do not significantly affect on another in this case.  

In [34]:
## 2. Find correlation between churn and return rate
# Create scatter plot
fig = px.scatter(rates, x='Return_Rate', y='Churn_Rate', 
                 trendline="ols", trendline_scope="overall")

# Calculate correlation coefficient
rates_corr = rates['Return_Rate'].corr(rates['Churn_Rate'])

# Add annotation for correlation coefficient
fig.add_annotation(x=41.5, y=20.15, 
                   text=f'R = {rates_corr:.3f}', 
                   showarrow=False, 
                   font=dict(size=13, color="red"))

fig.show()

This scatter plot shows a weak negative linear relationship between churn rate and return rate, suggesting that as one variable increases, the other tends to decrease slightly. However, the correlation coefficient is close to zero, indicating a very weak relationship and little evidence of correlation between the two variables.

### **5. Conclusion & Recommendation (SHARE)**

#### **⦿ Key Insights**

**Customer Demographics and Behaviors:**
* Spending behaviors are similar across genders, with males slightly outnumbering females.
* The dataset comprises a diverse customer base, primarily aged between 20 and 69.
* Customers aged 20-29 prefer Home goods, while those in the 40-49 age group favor Clothing.
* Payment method preferences are evenly distributed, with credit cards slightly prevailing.

**Product and Sales Trends:**
* Clothing is the most popular category, while Home generates the highest profits. Electronics has the highest return rate, whereas Books generate the lowest sales revenue.
* Sales revenue consistently hovers around $51M per year with minor fluctuations.
* February consistently records the lowest total sales, possibly due to fewer days, but not necessarily the lowest average monthly sales.
* The return rate has remained consistently high at around 40% since 2020, potentially indicating issues with product satisfaction or quality.

**Customer Segmentation based on CLV:**
* The medium segment holds the majority of customers and generates the highest revenue, followed closely by the high segment.
* Churn and return rates across all segments hover around 20% and 40%, respectively, with the medium segment exhibiting the highest rates.
* Age distributions are similar across segments, with the highest concentration observed among customers aged 20-29.
* The low segment shows a slightly higher usage of Cash, while the high segment exhibits a slightly higher preference for PayPal.

#### **⦿ Business & Marketing Recommendations**

**Targeted Marketing Strategies:**
* Tailor marketing campaigns to different customer segments based on their purchasing behavior, CLV, and demographic characteristics like gender and age.
  * Example: Create specific promotions for Home goods aimed at the 20-29 age group and Clothing for the 40-49 age group.
* Offer personalized promotions, discounts, and messaging based on segment characteristics and purchasing behaviors to improve customer engagement and loyalty.
* Introduce seasonal sales initiatives to offset slower or shorter months such as February and September.
  * Example: Offer special discounts or curated product sets for Valentine's Day, catering to the holiday gift market.

**Product & Quality Optimization:**
* Utilize sales data and trends to optimize inventory management practices, ensuring adequate stock levels of high-demand/profit products such as Home and Clothing, while minimizing excess inventory for products with lower sales volumes. 
* Identify and address the root causes of high return rates, particularly for Electronics, to enhance product quality and customer satisfaction.
* Implement bundling strategies to encourage larger purchases and increase revenue by offering complementary products at discounted rates.
  * Example: Provide discounted bundles for purchasing multiple books from the same series. 

**Customer Retention and Loyalty Programs:**
* Implement targeted customer retention initiatives such as loyalty programs or exclusive perks, to reduce churn rates and encourage repeat purchases. 
  * Example: Offer VVIP or member-exclusive benefits and rewards.
* Focus on improving overall customer satisfaction by addressing issues related to product quality, customer service, and delivery experience, which could contribute to the high return rates.

**Payment Method Optimization:**
* Offer incentives or discounts for using credit cards, considering their slight prevalence over other payment methods. Collaborate with banks to offer credit card exclusive reward points or cashback offers.
* Continue to enhance the payment experience by ensuring seamless integration of payment gateways and expand more digital payment options to cater to diverse customer preferences. 
  * Ex: offering intallment plans or accepting crypto-payments. 

**Performance Monitoring and Analysis:**
* Establish regular performance monitoring measurements to track key metrics(KPIs) such as sales revenue, churn rates, return rates, and customer satisfactions.
* Conduct regular quality checks and customer feedback surveys to address any issues promptly and continually improve overall customer experience.
* Perform ongoing analysis and market research to identify emerging trends, patterns, and opportunities for product improvement or innovation.